Failure to upgrade from EOL'd release

Following instructions as per https://support.system76.com/articles/upgrade-pop/#advanced-install-terminal:

```sh
pop-upgrade release upgrade systemd
```

fails due to missing `sudo` with:

```sh
/sys/firmware/dmi/tables/smbios_entry_point: Permission denied
/dev/mem: Permission denied
Event: removing deprecated and/or conflicting packages
Event: updating package lists for the current release
Release upgrade status: release upgrade aborted: failed to update package lists for the current release: command failed with exit status: exit code: 100
pop-upgrade: dbus client error: calling ReleaseUpgradeFinalize method failed
```

Running with sudo afterwards fails with:

```sh
Event: removing deprecated and/or conflicting packages
Event: updating package lists for the current release
Release upgrade status: release upgrade aborted: failed to update package lists for the current release: command failed with exit status: exit code: 100
pop-upgrade: dbus client error: calling ReleaseUpgradeFinalize method failed
```

In GNOME Settings it shows Ubuntu 19.04. I can see the Download button. Pressing it results in the same Error 100.

It also seems my `sources.list` has been messed up.

Output of `sudo apt update`:

```sh
Hit:1 http://packages.microsoft.com/repos/vscode stable InRelease
Hit:2 http://ppa.launchpad.net/linuxuprising/shutter/ubuntu disco InRelease                                     
Ign:3 http://dl.google.com/linux/chrome/deb stable InRelease                                                    
Ign:4 http://old-releases.ubuntu.com/ubuntu disco InRelease                                                                   
Hit:5 http://dl.google.com/linux/chrome/deb stable Release                                                                    
Ign:6 http://old-releases.ubuntu.com/ubuntu disco-updates InRelease                                                           
Hit:7 http://ppa.launchpad.net/obsproject/obs-studio/ubuntu disco InRelease                                                   
Ign:8 http://old-releases.ubuntu.com/ubuntu disco-security InRelease                     
Ign:9 http://old-releases.ubuntu.com/ubuntu disco-backports InRelease              
Hit:10 http://ppa.launchpad.net/samoilov-lex/sway/ubuntu disco InRelease
Err:11 http://old-releases.ubuntu.com/ubuntu disco Release          
  404  Not Found [IP: 2001:67c:1360:8001::25 80]
Err:13 http://old-releases.ubuntu.com/ubuntu disco-updates Release                 
  404  Not Found [IP: 2001:67c:1360:8001::25 80]
Hit:14 http://ppa.launchpad.net/system76/pop/ubuntu disco InRelease 
Get:15 https://gitlab.com/paulcarroty/vscodium-deb-rpm-repo/raw/repos/debs vscodium InRelease [3,828 B]  
Err:16 http://old-releases.ubuntu.com/ubuntu disco-security Release                                                    
  404  Not Found [IP: 2001:67c:1360:8001::25 80]
Hit:17 http://ppa.launchpad.net/wireguard/wireguard/ubuntu disco InRelease         
Err:18 http://old-releases.ubuntu.com/ubuntu disco-backports Release
  404  Not Found [IP: 2001:67c:1360:8001::25 80]
Reading package lists... Done
E: The repository 'http://old-releases.ubuntu.com/ubuntu disco Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://old-releases.ubuntu.com/ubuntu disco-updates Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://old-releases.ubuntu.com/ubuntu disco-security Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: The repository 'http://old-releases.ubuntu.com/ubuntu disco-backports Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```
and `cat /etc/apt/sources.list`:

```sh
# deb cdrom:[Pop_OS 18.04 _Bionic Beaver_ - Release amd64 (20180612)]/ bionic main restricted
deb-src http://old-releases.ubuntu.com/ubuntu/ disco main restricted universe multiverse
deb http://old-releases.ubuntu.com/ubuntu/ disco-updates main restricted universe multiverse
deb-src http://old-releases.ubuntu.com/ubuntu/ disco-updates main restricted universe multiverse
deb http://old-releases.ubuntu.com/ubuntu/ disco-security main restricted universe multiverse
deb-src http://old-releases.ubuntu.com/ubuntu/ disco-security main restricted universe multiverse
deb http://old-releases.ubuntu.com/ubuntu/ disco-backports main restricted universe multiverse
deb-src http://old-releases.ubuntu.com/ubuntu/ disco-backports main restricted universe multiverse
deb http://old-releases.ubuntu.com/ubuntu/ disco main restricted universe multiverse
# deb http://apt.pop-os.org/proprietary disco main
# deb-src http://apt.pop-os.org/proprietary bionic main

deb https://gitlab.com/paulcarroty/vscodium-deb-rpm-repo/raw/repos/debs/ vscodium main
```

Thanks, @mmstick, running that, followed by `sudo apt update` and `sudo apt full-upgrade` has got me a functional Pop!_OS 19.04 downgrade. Also, it got me past the libparted error in building the upgrade app, now make aborts with:

```sh
error[E0425]: cannot find function `head` in crate `reqwest`
   --> src/repair/sources.rs:145:14
    |
145 |     reqwest::head(&release_file).send().ok().map_or(false, |resp| resp.status().is_success())
    |              ^^^^ not found in `reqwest`

warning: unused `#[macro_use]` import
  --> src/lib.rs:13:1
   |
13 | #[macro_use]
   | ^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `reqwest::Client`
 --> src/repair/sources.rs:6:5
  |
6 | use reqwest::Client;
  |     ^^^^^^^^^^^^^^^

error: aborting due to previous error
```

--- a/src/repair/sources.rs
+++ b/src/repair/sources.rs
@@ -3,6 +3,7 @@
 use crate::release::eol::{EolDate, EolStatus};
 use apt_sources_lists::{SourceEntry, SourceError, SourcesLists};
 use distinst_chroot::Command;
+use reqwest::Client;
 use std::{fs, io, path::Path};
 use ubuntu_version::Codename;
 
@@ -40,7 +41,14 @@ pub fn repair(codename: Codename) -> Result<(), SourcesError> {
         // When EOL, the Ubuntu archives no longer carry packages for that release.
         // Also, disable the proprietary repository before upgrading an EOL release.
         sources_list.entries_mut(|entry| {
-            if replace_ubuntu_old_release(&mut entry.url) {
+            let url = entry.url();
+            if let Some(pos) = find_ubuntu_archive(url) {
+                let old_release = modify_to_old_release_archive(url, pos);
+
+                if release_exists(&entry, &old_release) {
+                    entry.url = old_release;
+                }
+
                 true
             } else if entry.url == "http://apt.pop-os.org/proprietary" {
                 entry.enabled = false;
@@ -125,14 +133,16 @@ fn create_new_sources_list(release: &str) -> io::Result<()> {
     Ok(())
 }
 
-fn replace_ubuntu_old_release(url: &mut String) -> bool {
-    if let Some(pos) = twoway::find_str(&url, "archive.ubuntu") {
-        let stripped = &url[pos + 7..];
-        *url = ["http://old-releases", stripped].concat();
-        return true;
-    }
+fn find_ubuntu_archive(url: &str) -> Option<usize> { twoway::find_str(&url, "archive.ubuntu") }
 
-    false
+fn modify_to_old_release_archive(url: &str, pos: usize) -> String {
+    let stripped = &url[pos + 7..];
+    ["http://old-releases", stripped].concat()
+}
+
+fn release_exists(entry: &SourceEntry, new_url: &str) -> bool {
+    let release_file = [new_url, "/dists/", &entry.suite, "/Release"].concat();
+    Client::new().head(&release_file).send().ok().map_or(false, |resp| resp.status().is_success())
 }
 
 #[cfg(test)]
@@ -142,7 +152,8 @@ mod tests {
     #[test]
     fn old_releases() {
         let mut url = String::from("http://us.archive.ubuntu.com/ubuntu/");
-        replace_ubuntu_old_release(&mut url);
+        let pos = find_ubuntu_archive(&url).unwrap();
+        url = modify_to_old_release_archive(&url, pos);
         assert_eq!(&url, "http://old-releases.ubuntu.com/ubuntu/");
     }
 }

